policyeval/base_worker.go

```diff
@@ -22,23 +24,75 @@ import (
 var errTargetNotReady = errors.New("target not ready")
 
 // Worker is responsible for executing a policy evaluation request.
-type Worker struct {
+type BaseWorker struct {
+	id            string
 	logger        hclog.Logger
 	pluginManager *manager.PluginManager
-	policyManager *Manager
+	policyManager *policy.Manager
+	broker        *Broker
+	queue         string
 }
 
-// NewWorker returns a new Worker instance.
-func NewWorker(l hclog.Logger, pm *manager.PluginManager, m *Manager) *Worker {
-	return &Worker{
-		logger:        l.Named("worker"),
+// NewBaseWorker returns a new BaseWorker instance.
+func NewBaseWorker(l hclog.Logger, pm *manager.PluginManager, m *policy.Manager, b *Broker, queue string) *BaseWorker {
+	id := uuid.Generate()
+
+	return &BaseWorker{
+		id:            id,
+		logger:        l.Named("worker").With("id", id, "queue", queue),
 		pluginManager: pm,
 		policyManager: m,
+		broker:        b,
+		queue:         queue,
+	}
+}
+
+func (w *BaseWorker) Run(ctx context.Context) {
+	w.logger.Debug("starting worker")
+
+	for {
+		select {
+		case <-ctx.Done():
+			w.logger.Info("stopping worker")
+			return
+		default:
+		}
+
+		eval, token, err := w.broker.Dequeue(ctx, w.queue)
+		if err != nil {
+			w.logger.Warn("failed to dequeue evaluation", "err", err)
+			continue
+		}
+
+		if eval == nil {
+			// Nothing to do for now or we timedout, let's loop.
+			continue
+		}
+
+		logger := w.logger.With(
+			"eval_id", eval.ID,
+			"eval_token", token,
+			"policy_id", eval.Policy.ID)
+
+		if err := w.handlePolicy(ctx, eval); err != nil {
+			logger.Error("failed to evaluate policy", "err", err)
+
+			// Notify broker that policy eval was not successful.
+			if err := w.broker.Nack(eval.ID, token); err != nil {
+				logger.Warn("failed to NACK policy evaluation", "err", err)
+			}
+			continue
+		}
+
+		// Notify broker that policy eval was successful.
+		if err := w.broker.Ack(eval.ID, token); err != nil {
+			logger.Warn("failed to ACK policy evaluation", "err", err)
```

Comment:
Yes, if the ACK fails, the broker will re-try this. But if this scaling action succeeded, the next eval will no-op because the target is already scaled. 

Also there aren't many ways the ACK would fail. I copied this from Nomad an there this call happens over RPC, so network is an issue there.